File: src/main/java/rustleund/fightingfantasy/framework/closures/impl/ChainedClosure.java
package rustleund.fightingfantasy.framework.closures.impl;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import rustleund.fightingfantasy.framework.base.GameState;
import rustleund.fightingfantasy.framework.closures.Closure;

public class ChainedClosure implements Closure {

	private List<Closure> closures;

	public ChainedClosure() {
		this.closures = new ArrayList<Closure>();
	}

	public ChainedClosure(List<Closure> closures) {
		this.closures = closures;
	}

	public ChainedClosure(Closure... closures) {
		this.closures = new ArrayList<Closure>(Arrays.asList(closures));
	}

	public void addClosure(Closure closure) {
		this.closures.add(closure);
	}

	@Override
	public boolean execute(GameState gameState) {
		if (this.closures != null) {
			for (Closure closure : this.closures) {
				if (!closure.execute(gameState)) {
					return false;
				}
			}
		}
		return true;
	}

	@Override
	public String toString() {
		return this.closures.toString();
	}

}

File: src/main/java/rustleund/fightingfantasy/framework/closures/impl/ChainedClosure.kt
package rustleund.fightingfantasy.framework.closures.impl

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonSerializationContext
import rustleund.fightingfantasy.framework.base.GameState
import rustleund.fightingfantasy.framework.closures.Closure
import rustleund.fightingfantasy.gamesave.SerializableClosure

class ChainedClosure(private val closures: List<Closure> = listOf()) : SerializableClosure {

    override fun execute(gameState: GameState) = closures.all { it.execute(gameState) }

    override fun serialize(context: JsonSerializationContext): JsonElement {
        val result = JsonObject()
        result.add("closures", context.serialize(closures))
        return result
    }

    override fun toString(): String {
        return closures.toString()
    }
}